integrations/google-analytics-4/lib/index.js

```diff
@@ -0,0 +1,303 @@
+   *   - By User ID, Google signals, then device (userIdSignalsAndDevice)
+   *   - By User ID and Devicea (userIdAndDevice)
+   * 
+   * Google's Reporting Identity: https://support.google.com/analytics/answer/9213390?hl=en
+   * 
+   * Note that the user ID can be appended as part of the user_properties
+   * object instead of being configured by an explicit command.
+   * https://developers.google.com/analytics/devguides/collection/ga4/cookies-user-id#set_user_id
+   */
+  var userId = identify.userId();
+  var validReportingIdentity = opts.googleReportingIdentity === 'userIdSignalsAndDevice' || opts.googleReportingIdentity === 'userIdAndDevice'
+  if (userId && validReportingIdentity) {
+    userProperties.user_id = userId;
+  }
+
+  if (Object.keys(userProperties).length) {
+    window.gtag('set', 'user_properties', userProperties);
+  }
+};
+
+/**
+ * Group
+ *
+ * @api public
+ * @param {Facade.Group} group
+ */
+GA4.prototype.group = function(group) {
+  window.gtag('event', 'join_group', {
+    group_id: group.groupId()
+  });
+};
+
+/**
+ * Page
+ *
+ * @api public
+ * @param {Facade.Page} page
+ */
+GA4.prototype.page = function(page) {
+  // If the Send Google's Automatic Page View Measurement setting is set to true then
+  // don't handle page calls to avoid duplicate page_view events.
+  if (this.options.sendAutomaticPageViewEvent) {
+    return;
+  }
+
+  // Track categorized pages setting needed here?
+
+  var props = page.properties();
+  var name = page.fullName();
+
+  var pageLocation = props.url;
+  var pageReferrer = page.referrer();
+  var pageTitle = name || props.title;
+
+  window.gtag('event', 'page_view', {
+    page_location: pageLocation,
+    page_referrer: pageReferrer,
+    page_title: pageTitle
+  });
+};
+
+/**
+ * Track
+ *
+ * @api public
+ * @param {Track} track
+ */
+
+GA4.prototype.track = function(track) {
+
+  var mappings = this.options.customEventsAndParameters;
+
+  for (var i = 0; i < mappings.length; i++) {
+    var mapping = mappings[i]; // Type check for object?
+    var segmentEvent = mapping.segmentEvent;
+    var googleEvent = mapping.googleEvent;
+  
+    if (!segmentEvent || !googleEvent || segmentEvent !== track.event()) {
+      continue;
+    }
+
+    var parameterMappings = mapping.parameters || []; // Type check for array?
+    var parameters = {};
+
+    // Map Segment event fields to Google Event Parameters.
+    // Text map settings that are nested in a mixed settings take on a different shape
+    // than a top-level text map setting.
+    // eg; [{ key: 'properties.genre', value: 'primary_genre }]
+    //
+    for (var j = 0; j < parameterMappings.length; j++) {
+      var map = parameterMappings[j] || {}; // Type check for object?
+      if (!map.key || !map.value) {
+        continue;
+      }
+
+      var param = map.value;
+      var value = track.proxy(map.key);
+      parameters[param] = value;
+    }
+
+    window.gtag('event', googleEvent, reject(parameters));
```

Comment:
Wondering if we need `reject` here given line 292 above - if `map. value` is falsy, then we don't add the key:value pair to the parameters map in the first place.

Reply:
Good call - it's pretty redundant!